sdcm/utils/common.py

```diff
@@ -429,53 +427,80 @@ def inner(*args, **kwargs):
-                elif isinstance(obj, dict):
-                    futures.append(pool.submit(func, **obj))
-                else:
-                    futures.append(pool.submit(func, obj))
-
-            for future in futures:
-                try:
-                    result = future.result(self.timeout)
-                    results.append(FutureResult(exc=None, result=result))
-                except Exception as ex:  # pylint disable=broad-except
-                    LOGGER.warning('Error happened during running %s in %s:\n%s',
-                                   func.__name__,
-                                   future,
-                                   ex.__class__.__name__)
-                    if ignore_exceptions:
-                        results.append(FutureResult(exc=ex, result=None))
-                        continue
-                    raise
+            func = func_wrap(func)
+
+        futures = []
+
+        for obj in self.objects:
+            if unpack_objects and isinstance(obj, (list, tuple)):
+                futures.append(self._thread_pool.submit(func, *obj))
+            elif unpack_objects and isinstance(obj, dict):
+                futures.append(self._thread_pool.submit(func, **obj))
+            else:
+                futures.append(self._thread_pool.submit(func, obj))
+        time_out = self.timeout
+        for obj_idx, future in enumerate(futures):
+            try:
+                result = future.result(time_out)
+            except FuturesTimeoutError as exception:
+                results.append(ParallelObjectResult(obj=self.objects[obj_idx], exc=exception, result=None))
+                time_out = 0.001  # if there was a timeout on one of the futures there is no need to wait for all
+            except Exception as exception:  # pylint: disable=broad-except
+                results.append(ParallelObjectResult(obj=self.objects[obj_idx], exc=exception, result=None))
+            else:
+                results.append(ParallelObjectResult(obj=self.objects[obj_idx], exc=None, result=result))
+
+        self.clean_up(futures)
+
+        if ignore_exceptions:
+            return results
+
+        timed_out = [result for result in results if isinstance(result.exc, FuturesTimeoutError)]
+        if timed_out:
+            raise FuturesTimeoutError("when running on: %s" % [r.obj for r in results])
+        runs_that_finished_with_exception = [res for res in results if res.exc]
+        if runs_that_finished_with_exception:
+            raise ParallelObjectException(results=results)
         return results
 
+    def clean_up(self, futures):
+        # if there are futures that didn't run  we cancel them
+        for future in futures:
+            future.cancel()
+        self._thread_pool.shutdown(wait=False)
+        # we need to unregister internal function that waits for all threads to finish when interpreter exits
+        from concurrent.futures.thread import _python_exit
```

Comment:
can we move this to the start of the file ?

Reply:
moved